File: core/src/main/java/org/elkd/core/log/Log.java
package org.elkd.core.log;

import java.util.List;

public interface Log<T> {

  long INDEX_FROM = -1;

  long append(T entry);

  long append(long index, T entry);

  T read(long index);

  List<T> read(long from, long to);

  CommitResult<T> commit(long index);

  void revert(long index);

  long getCommitIndex();

  long getLastIndex();
}

File: core/src/main/java/org/elkd/core/log/Log.kt
package org.elkd.core.log

interface Log<E : LogEntry> {
  val commitIndex: Long
  val lastIndex: Long
  val lastEntry: E

  fun append(entry: E): Long

  fun append(index: Long, entry: E): Long

  fun read(index: Long): E?

  fun read(from: Long, to: Long): List<E>

  fun commit(index: Long): CommitResult<E>

  fun revert(index: Long)
}
